In [ ]:
사이킷런으로 시작하는 머신러닝
1. 사이킷런 소개와 특징

In [2]:
import sklearn
print(sklearn.__version__)

0.24.2


In [ ]:
2. 첫 번째 머신러닝 만들어 보기 - 붓꽃 품종 예측하기

- 붓꽃 data set : 꽃잎의 길이, 너비, 꽃받침의 길이, 너비
- 분류는 대표적인 지도학습 방법의 하나이다.
- 지도학습은 학습을 위한 다양한 피처와 분류 결정값인 레이블 데이터로 모델을 학습한 뒤, 별도의 테스트 데이터 세트에서 미지의 레이블을 예측한다.
- 학습 데이터 세트 : 학습을 위해 주어진 데이터 세트
- 테스트 데이터 세트 : 머신러닝 모델의 예측 성능을 평가하기 위해 별도로 주어진 데이터 세트
- 사이킷런 패키지 내 모듈명은 sklearn 으로 시작하는 명명규칙이 있다.
- sklearn.tree 모듈 : 트리 기반 ml 알고리즘을 구현한 클래스의 모임
- sklearn.medel_selection : 학습 데이터와 검증 데이터, 예츨 데이터로 데이터를 분리하거나 최적의 하이퍼 파라미터로 평가하기 위한 모듈의 모임.
- 하이퍼 파라미터 : 머신러닝 알고리즘별로 최적의 학습을 위해 직접 입력하는 파라미터들. 이를 통해 머신러닝 성능을 튜닝할 수 있다.

In [3]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

- load_iris() : 붓꽃 데이터 세트를 생성

In [5]:
import pandas as pd

# 붓꽃 데이터 세트를 로딩한다.
iris = load_iris()

# iris.data 는 Iris 데이터 세트에서 피처(feature)만으로 된 데이터를 numpy로 가지고 있습니다.
iris_data = iris.data

# iris.target 은 붓꽃 데이터 세트에서 레이블(결정 값) 데이터를 numpy로 가지고 있습니다.
iris_label = iris.target #setosa, vesicolor, virginica 3 종류가 있고, 각각 0,1,2 로 되어 있다는 뜻.
print('iris target값 :', iris_label)
print('iris target명 :', iris.target_names)

# 붓꽃 데이터 세트를 자세히 보기 위해 DataFrame 으로 변환합니다.
iris_df = pd.DataFrame(data=iris_data, columns=iris.feature_names)
iris_df['label']=iris.target
iris_df

iris target값 : [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
iris target명 : ['setosa' 'versicolor' 'virginica']


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     label  
0        0  
1        0  
2        0  
3        0  
4        0  
..     ...  
145      2  
146      2  
147      2  
148      2  
149      2  

[150 rows x 5 columns]

<학습용 데이터와 테스트용 데이터 분리>
- 학습용 데이터와 테스트용 데이터는 반드시 분리해야 함.
- train_test_split() : 학습 데이터와 테스트 데이터를 test_size 파라미터 입력 값의 비율로 쉽게 분할한다. ex) test_size=0.2 로 하면, 전체 데이터 중 테스트 데이터가 20%, 학습 데이터가 80%로 데이터를 분할한다.

In [9]:
# 학습용 데이터와 테스트용 데이터 분리
# 학습용 피처 데이터 세트를 X_train 이고, 테스트 용 피처 데이터 세트를 X_test로, 학습용 레이블 데이터 세트를 y_train으로, 테스트용 레이블 데이터 세트를 y_test 로 반환
X_train, X_test, y_train, y_test = train_test_split(iris_data, iris_label, test_size=0.2, random_state=11)

- train_test_split() 의 첫 번째 파라미터 인, iris_data 는 피처 데이터 세트.
- 두 번째 파라미터 iris_label 은 레이블 데이터 세트
- 전체 데이터 세트 중 테스트 데이터 세트의 비율
- random_state 는 호출할 때 마다 같은 학습/테스트 용 데이터 세트를 생성하기 위해 주어진 난수 발생 값. ==> 지정하지 않으면 수행할 때 마다 다른 학습/ 테스트 용 데이터를 만듦. (숫자는 뭐든 상관 없음)

In [10]:
# DecisionTreeClassifier(사이킷런의 의사 결정 트리 클래스) 객체 생성
dt_clf = DecisionTreeClassifier(random_state=11)

- 생성된 DecisionTreeClassifier 객체의 fit() 메서드에 학습용 피처 데이터 속성과 결정값 데이터 세트를 입력해 호출하면 학습을 수행한다.
- 아래의 코드가 수행되면 : 의사 결정 트리 기반의 DecisionTreeClassifier 객체는 학습 데이터를 기반으로 학습이 완료 ==> 이를 기반으로 예측 수행 가능. 예측은 반드시 학습 데이터가 아닌 다른 데이터를 이용해야 함. 일반적으로 테스트 데이터 세트를 이용

In [11]:
# 모델 학습
# 학습 수행
dt_clf.fit(X_train, y_train)

DecisionTreeClassifier(random_state=11)

- DecisonTreeClassifier 객체의 predict() 메서드에 테스트용 피처 데이터 세트를 입력해 호출하면 학습된 모델 기반에서 데스트 데이터 세트에 대한 예측값을 반환

In [12]:
# 예측 수행
# 학습이 완료된 DecisionTreeClassifier 객체에서 테스트 데이터 세트로 예측 수행.
pred = dt_clf.predict(X_test)

- 예측 결과를 기반으로 의사결정트리 기반의 DecisionTreeClassifier의 예측 성능을 평가.
- 일반적으로 머신러닝 모델의 성능 평가 방법은 여러가지가 있다.
- 아래에선 정확도를 측정해보겠다.
- 정확도 : 예측 결과가 실제 레이블 값과 얼마나 정확하게 맞는지를 평가하는 지표
- accuracy_score() : 사이킷런의 정확도 측정 함수
- accuracy_score 의 첫 번째 파라미터는 실제 레이블 데이터 세트, 두 번째 파라미터로 예측 레이블 데이터 세트 입력

In [13]:
# 정확도 측정
from sklearn.metrics import accuracy_score
print('예측 정확도 : {0:4f}'.format(accuracy_score(y_test,pred)))

예측 정확도 : 0.933333


<붓꽃 데이터 분류 예측 프로세스>
1. 데이터 세트 분리 : 데이터를 학습 데이터와 테스트 데이터로 분리
2. 모델 학습 : 학습 데이터를 기반으로 ML 알고리즘을 적용해 모델을 학습시킴
3. 예측 수행 : 학습된 ML 모델을 이용해 테스트 데이터의 분류(붓꽃 종류) 를 예측한다.
4. 평가 : 이렇게 예측된 결괏값과 테스트 데이터의 실제 결괏값을 비교해 ML 모델 성능을 평가

In [ ]:
3. 사이킷런의 기반 프레임워크 익히기

- 사이킷런은 ML 모델 학습을 위해 : fit() 을
- 학습된 모델의 예측을 위해 : predict() 매서드를 제공.
- 지도 학습의 주요 두 축인 분류와 회귀의 다양한 알고리즘을 구현한 모든 사이킷런 클래스는 fit()과 predict() 만을 이용해 간단하게 학습과 예측 결과를 반환
- 분류 알고리즘을 구현한 클래스 : Classifier로,
- 회귀 알고리즘을 구현한 클래스 : Regressor 로 지칭
- Estimator : Classifier 과 Regaressor 을 합침 (당연히 내부에서 fit(), predict() 구현)

<사이킷런의 주요 모듈>
1. 예제 데이터
1) sklearn.datasets : 사이킷런에 내장되어 예제로 제공하는 데이터 세트
2. 피처 처리
1) sklearn.preprocessing : 데이터 전처리에 필요한 다양한 가공 기능 제공 (문자열을 숫자형 코드 값으로 인코딩, 정규화, 스케일링 등)
2) sklearn.feature_selection : 알고리즘에 큰 영향을 미치는 피처를 우선순위대로 셀렉션 작업을 수행하는 다양한 기능 제공
3) sklearn.feature_extraction : 텍스트 데이터나 이미지 데이터의 벡터화된 피처를 추출하는데 사용됨.
3. 피처 처리 & 차원 축소
1) sklearn.decomposition : 차원 축소와 관련한 알고리즘을 지원하는 모듈.
4. 데이터 분리, 검증&파라미터 튜닝
1) sklearn.model_selection : 교차 검증을 위한 학습용/테스트용 분리
5. 평가
1) sklearn.metrics : 분류, 회귀, 클러스터링, 페어와이즈에 대한 다양한 성능 측정 방법 제공
6. ML 알고리즘
1) sklearn.ensemble : 앙상블 알고리즘 제공
2) sklearn.linear_model : 주로 선형회귀, 릿지, 라쏘 및 로지스틱 회귀 등 회귀 관련 알고리즘을 지원
3) sklearn.naive_bayes : 나이브 베이즈 알고리즘 제공. 가우시안, NB, 다항분포 등
4) sklearn.neighbors : 최근접 이웃 알고리즘 제공
5) sklearn.svm : 서포트 벡터 머신 알고리즘 제공
6) sklearn.tree : 의사 결정 트리 알고리즘 제공
7) sklearn.cluster : 비지도 클러스터링 알고리즘 제공
7. 유틸리티
1) sklearn.pipline : 피처 처리 등의 변환과 ML 알고리즘 학습, 예측등을 함께 묶어 실행가능한 유틸리티 제공

일반적으로 머신러닝 모델을 구축하는 주요 프로세스는 피처의 가공, 변경, 추출을 수행하는 피처 처리, ML 알고리즘 학습/ 예측 수행, 그리고 모델 평가의 단계를 반복적으로 수행하는 것

<내장된 예제 데이터 세트>
두가지로 나뉨. 대부분 일반적으로 '딕셔너리' 형태로 저장되이 있음.
1. 분류나 회귀를 연습하기 위한 예제 용도
2. 분류나 클러스터링을 위해 표본 데이터로 생성될 수 있는 데이터 세트
3. 예제 데이터로는 : datasets.load_boston(), ~.load_breat_cancer(), ~.load_diabetes(), ~.load_digits(), ~.load_iris(), ~.make_classifications(), ~.make_blobs() 등이 있음.
4. fetch 계열의 명령은 데이터가 커서 패키지에 저장되어 있지 않고, 인터넷에 내려받아 사용함.

<데이터 세트의 딕셔너리로 저장된 키 의 의미>
1. data : 피처의 데이터 세트를 가리킴 ==> ndarray 타입
2. target : 분류 시 레이블 값, 회귀일 땐 숫자 결괏값 데이터 세트 ==> ndarray 타입
3. target_name : 개별 레이블의 이름을 나타냄 ==> list 타입
4. feature_names : 피처의 이름을 나타냄 ==> list 타입
5. DESCR : 데이터 세트에 대한 설명과 각 피처의 설명을 나타냄 ==> 스트링 타입

In [14]:
from sklearn.datasets import load_iris

iris_data = load_iris()
print(type(iris_data)) # 딕셔너리 형태를 가짐

<class 'sklearn.utils.Bunch'>


In [15]:
#딕셔너리 형태로 저장된 데이터 세트의 key 출력
keys = iris_data.keys()
print('붓꽃 데이터 세트의 키들:', keys)

붓꽃 데이터 세트의 키들: dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])


- 데이터 키는 피처들의 데이터 값을 가리킵니다.
- 데이터 세트가 딕셔너리 형태이기 때문에 피처 데이터 값을 추출하기 위해서는 '데이터세트.data' (혹은 데이터세트['data'] ) 를 이용하면 됨. (위의 코드에서 출력한 data, target, frame .. 등 모두)

In [19]:
print('\n feature_names 의 type:', type(iris_data.feature_names))
print('feature_names 의 shape:', len(iris_data.feature_names))
print(iris_data.feature_names) # key 이기 때문에 딕셔너리.key 혹은 딕셔너리[key] 로 출력가능

print('\n target_names 의 type:', type(iris_data.target_names))
print('target_names 의 shape:', len(iris_data.target_names))
print(iris_data.target_names)

print('\n data 의 type:', type(iris_data.data))
print('data 의 shape:', iris_data.data.shape)
print(iris_data['data'])

print('\n target 의 type:', type(iris_data.target))
print('target 의 shape:', len(iris_data.target))
print(iris_data['target'])


 feature_names 의 type: <class 'list'>
feature_names 의 shape: 4
['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']

 target_names 의 type: <class 'numpy.ndarray'>
target_names 의 shape: 3
['setosa' 'versicolor' 'virginica']

 data 의 type: <class 'numpy.ndarray'>
data 의 shape: (150, 4)
[[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]
 [5.4 3.7 1.5 0.2]
 [4.8 3.4 1.6 0.2]
 [4.8 3.  1.4 0.1]
 [4.3 3.  1.1 0.1]
 [5.8 4.  1.2 0.2]
 [5.7 4.4 1.5 0.4]
 [5.4 3.9 1.3 0.4]
 [5.1 3.5 1.4 0.3]
 [5.7 3.8 1.7 0.3]
 [5.1 3.8 1.5 0.3]
 [5.4 3.4 1.7 0.2]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.  0.2]
 [5.1 3.3 1.7 0.5]
 [4.8 3.4 1.9 0.2]
 [5.  3.  1.6 0.2]
 [5.  3.4 1.6 0.4]
 [5.2 3.5 1.5 0.2]
 [5.2 3.4 1.4 0.2]
 [4.7 3.2 1.6 0.2]
 [4.8 3.1 1.6 0.2]
 [5.4 3.4 1.5 0.4]
 [5.2 4.1 1.5 0.1]
 [5.5 4.2 1.4 0.2]
 [4.9 3.1 1.5 0.2]
 [5.  3.2 1.2 0.2]


4. Model Selection 모듈 소개
- 사이킷런의 model_selection 모듈 : 학습 데이터와 테스트 데이터 세트를 분리하거나 교차 검증 분할 및 평가, 그리고 Estimator의 하이퍼 파라미터를 튜닝하기 위한 다양한 함수와 클래스를 제공

<학습 / 테스트 데이터 세트 분리 - train_test_split()>

In [20]:
# 테스트 데이터 세트를 이용하지 않고, 학습 데이터 세트로만 학습하고 예측하면 무엇이 문제인지 다음을 통해 볼 수 있음.
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

iris = load_iris()
dt_clf = DecisionTreeClassifier()
train_data = iris.data
train_label = iris.target
#학습 수행
dt_clf.fit(train_data, train_label)

#학습 데이터 세트으로 예측 수행
pred=dt_clf.predict(train_data)
print('예측 정확도 :', accuracy_score(train_label, pred))

예측 정확도 : 1.0


- 정확도가 100% 이라 이상함. 이렇게 나온 이유는, 이미 학습한 학습 데이터 세트를 기반으로 예측했기 때문이다.
- 따라서 예측을 수행하는 데이터 세트는, 학습을 수행한 학습용 데이터 세트가 아닌 전용의 테스트 데이터 세트여야 함.
- 사이킷런의 train_test_split() 를 통해 원본 데이터 세트에서 학습 및 테스트 데이터 세트를 쉽게 분리할 수 있다.

<train_test_split() 이용해 붓꽃 데이터 세트를 학습 및 테스트>
1. sklearn.model_selection 모듈에서 train_test_split() 을 로드한다.
2. train_test_split() 는 첫 번째 파라미터로 피처 데이터 세트, 두 번째 파라미터로 레이블 데이터 세트를 입력받는다.
3. 그리고 선택적으로 다음 파라미터를 입력받음
1) test_size : 전체 데이터에서 테스트 데이터 세트 크기를 얼마로 샘플링할 것인지 결정. 디폴트는 0.25 즉 25%
2) train_size : 전체 데이터에서 학습용 데이터 세트 크기를 얼마로 샘플링할 것인가 결정. 잘 사용하지는 않음
3) shuffle : 데이터를 분리하기 전에 데이터를 미리 섞을지 결정. 디폴트는 TRUE. 데이터를 분산시켜서 좀 더 효율적인 학습 및 테스트 데이터 세트를 만드는데 사용한다.
4) random_state : random_state 는 호출할 때마다 동일한 학습/ 테스트용 데이터 세트를 생성하기 위해 주어지는 난수 값이다. train_test_split() 는 호출 시 무작위로 데이터를 분리하므로 random_state를 지정하지 않으면, 수행할 때마다 다른 학습/테스트 용 데이터를 생성한다. 이 책의 예제에서는 수행 할 때마다 동일한 데이터 세트로 분리하기 위해 random_state를 일정한 숫자 값으로 부여.
4. tarin_test_split() 의 반환값은 튜플 형태이다. 순차적으로 학습용 데이터의 피처 데이터 세트, 테스트용 데이터의 피처 데이터 세트, 학습용 데이터의 레이블 데이터 세트, 테스트용 데이터의 레이블 데이터 세트가 반환

In [1]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

dt_clf=DecisionTreeClassifier()
iris_data=load_iris()

X_train, X_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target, test_size=0.3, random_state=121)

In [2]:
dt_clf.fit(X_train, y_train)
pred=dt_clf.predict(X_test)
print('예측 정확도:{0:.4f}'.format(accuracy_score(y_test, pred)))

예측 정확도:0.9556


테스트 데이터로 예측을 수행한 결과 정확도가 약 95.65% 이다.

<교차 검증>
- 앞에서 알고리즘을 학습시키는 학습 데이터와 이에 대한 예측 성능을 평가하기 위한 별도의 테스트용 데이터가 필요하다고 말했다.
- 그러나 이 방법 역시 과적합에 취약한 약점을 가질 수 있다.
- 과적합 : 모델이 학습 데이터에만 과도하게 최적화되어, 실제 예측을 다른 데이터로 수행할 경우에는 예측 성능이 과도하게 떨어지는 것을 말함.
- 고정된 학습 데이터와 테스트 데이터로 평가를 하다보면, 테스트 데이터에만 최적의 성능을 발휘할 수 있도록 편향되게 모델을 유도하는 경향이 생기게 됨 ==> 해당 테스트 데이터에만 과적합되는 학습 모델이 만들어져 다른 테스트용 데이터가 들어올 경우 성능이 저하됨.
- 따라서 교차 검증을 이용해 더 다양한 학습과 평가를 수행
- 교차 검증 : 데이터 편중을 막기 위해, 별도의 여러 세트로 구성된 학습 데이터 세트와 검증 데이터 세트에서 학습과 평가를 수행하는 것
- 그리고 각 세트에서 수행한 평가 결과에 따라 하이퍼 파라미터 튜닝 등의 모델 최적화를 더욱 손쉽게 할 수 있다.
- 대부분의 ML 모델의 성능 평가는 교차 검증 기반으로 1차 평가를 한 뒤에 최종적으로 테스트 데이터 세트에 적용해 평가하는 프로세스.
- ML 에 사용되는 데이터 세트를 세분화해서 학습, 검증, 테스트 데이터 세트로 나눌 수 있습니다.
- 테스트 데이터 세트 외에 별도의 검증 데이터 세트를 둬서 최종 평가 이전에 학습된 모델을 다양하게 평가하는데 사용

<K 폴드 교차 검증>
- 가장 보편적으로 사용되는 교차 검증 기법
- k 폴드 교차 검증 : 먼저 k 개의 데이터 폴드 세트를 만들어서 k번만큼 각 폴드 세트에 학습과 검증 평가를 반복적으로 수행하는 방법
- 예)
1. 데이터 세트를 k 등분 한다
2. 첫 번째 반복에서는 처음부터 4등분을 하습 데이터 세트, 마지막 5번째 등분 하나를 검증 데이터 세트로 설정하고 학습 데이터 세트에서 학습 수행, 검증 데이터 세트에서 평가를 수행한다.
3. 그다음엔 5등분의 4번째를 다음엔 3번째....
- 이렇게 학습 데이터 세트와 검증 데이터 세트를 점진적으로 변경하면서 마지막 k 번째 까지 학습과 검증을 수행하는 것이 k 폴드 교차 검증.
- k 개의 예측 평가를 구했으면, 이를 평균해서 k 폴드 평가 결과로 반영하면 됨.
- 사이킷런에서는 k 폴드 교차 검증 프로세스를 구현하기 위해, kfold와 stratifiedkfold 클래스를 제공한다.

In [4]:
# KFold 클래스를 이용해 붓꽃 데이터 세트를 교차 검증하고 예측 정확도를 알아봄.
# 붓꽃 데이터 세트와 DecisionTreeClassifier를 다시 생성
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
import numpy as np

iris = load_iris()
features = iris.data
label = iris.target
dt_clf = DecisionTreeClassifier(random_state=156)

# 5개의 폴드 세트로 분리하는 KFold 객체와 폴드 세트별 정확도를 담을 리스트 객체 생성.
kfold = KFold(n_splits=5) #KFold 객체 생성 후 5개의 폴드 데이터 세트로 분리
cv_accuracy = []
print('붓꽃 데이터 세트 크기 :', features.shape[0])

붓꽃 데이터 세트 크기 : 150


- 전체 붓꽃 데이터는 150개 이므로, 학습용 데이터 세트는 이 중 4/5 개인 120개, 검증 테스트 데이터 세트는 1/5개인 30개로 분할
- KFold 객체는 split()을 호출하면, 학습용/검증용 데이터로 분할할 수 있는 인덱스를 반환
- 실제로 학습용/검증용 데이터 추출은 반환된 인덱스를 기반으로 개발 코드에서 직접 수행해야함.

In [5]:
# 5개의 폴드 세트를 생성하는 KFold 객체의 split()을 호출해 교차 검증 수행 시마다 학습과 검증을 반복해 예측 정확도를 측정.
n_iter=0

#KFold 객체의 split()를 호출하면 폴드 별 학습용, 검증용 테스트의 로우 인덱스를 array로 반환
for train_index, test_index in kfold.split(features):
    #kfold.split()으로 반환된 인덱스를 이용해 학습용, 검증용 테스트 데이터 추출
    X_train, X_test = features[train_index], features[test_index]
    y_train, y_test = label[train_index], label[test_index]
    #학습 및 예측
    dt_clf.fit(X_train, y_train)
    pred=dt_clf.predict(X_test)
    n_iter+=1
    #반복 시마다 정확도 측정
    accuracy = np.round(accuracy_score(y_test, pred), 4)
    train_size = X_train.shape[0]
    test_size = X_test.shape[0]
    print('\n#{0} 교차 검증 정확도 : {1}, 학습 데이터 크기 : {2}, 검증 데이터 크기: {3}'.format(n_iter, accuracy, train_size, test_size))
    print('#{0} 검증 세트 인덱스 :{1}'.format(n_iter, test_index))
    cv_accuracy.append(accuracy)

# 개별 iteration별 정확도를 합하여 평균 정확도 계산
print('\n## 평균 검증 정확도 :', np.mean(cv_accuracy))


#1 교차 검증 정확도 : 1.0, 학습 데이터 크기 : 120, 검증 데이터 크기: 30
#1 검증 세트 인덱스 :[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29]

#2 교차 검증 정확도 : 0.9667, 학습 데이터 크기 : 120, 검증 데이터 크기: 30
#2 검증 세트 인덱스 :[30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53
 54 55 56 57 58 59]

#3 교차 검증 정확도 : 0.8667, 학습 데이터 크기 : 120, 검증 데이터 크기: 30
#3 검증 세트 인덱스 :[60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83
 84 85 86 87 88 89]

#4 교차 검증 정확도 : 0.9333, 학습 데이터 크기 : 120, 검증 데이터 크기: 30
#4 검증 세트 인덱스 :[ 90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119]

#5 교차 검증 정확도 : 0.7333, 학습 데이터 크기 : 120, 검증 데이터 크기: 30
#5 검증 세트 인덱스 :[120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137
 138 139 140 141 142 143 144 145 146 147 148 149]

## 평균 검증 정확도 : 0.9


<Stratified K 폴드>
- Stratified K 폴드 : 이 폴드는 불균형한 분포도를 가진 레이블(결정 클래스) 데이터 집합을 위한 k 폴드 방식이다.
- 불균형한 분포도를 가진 레이블 데이터 집합 : 특정 레이블 값이 특이하게 많거나 매우 적어서 값의 불포가 한쪽으로 치우치는 것
- 예시) 대출 사기 데이터를 예측한다고 가정 => 데이터 세트 1억개, 대출 사기 여부를 뜻하는 레이블 : 1,0, 전체의 0.00001% 확률로 대출사기 레이블 존재. 이렇게 작은 비율로 1 레이블 값이 있다면, k 폴드로 랜덤하게 학습 및 테스트 세트의 인덱스를 고르더라도 레이블 값인 0,1 비율을 제대로 못 반영할 수 있음. 따라서 원본 데이터와 유사한 대출 사기 레이블 값의 불포를 학습/테스트 세트에도 유지하는게 매우 중요.
- Stratified K 폴드는 이처럼 K 폴드가 레이블 데이터 집합이 원본 데이터 집합의 레이블 분포를 학습 및 테스트 세트에 제대로 분배하지 못하는 경우의 문제를 해결해줌.
- Stratified K 폴드는 원본 데이터의 레이블 분포를 먼저 고려한 뒤, 이 분포와 동일하게 학습과 검증 데이터 세트를 분배.

In [12]:
#K 폴드가 가진 문제점 확인
import pandas as pd

iris=load_iris()
iris_df= pd.DataFrame(data=iris.data, columns=iris.feature_names) #붓꽃 데이터 세트를 간단하게 Dataframe 으로 생성하고, 레이블 값의 분포도를 확인
iris_df['label']=iris.target
iris_df['label'].value_counts()

2    50
1    50
0    50
Name: label, dtype: int64

- 레이블 값은 0,1,2 값 모두 50개로 동일.
- 이슈가 발생하는 현상을 도출하기 위해 3개의 폴드 세트를 KFold 로 생성하고, 각 교차 검증 시마다 생성되는 학습/검증 레이블 데이터 값의 분포도를 확인해보자

In [14]:
kfold = KFold(n_splits=3)
n_iter=0
for train_index, test_index in kfold.split(iris_df):
    n_iter +=1
    label_train=iris_df['label'].iloc[train_index]
    label_test=iris_df['label'].iloc[test_index]
    print('## 교차 검증 : {0}'.format(n_iter))
    print('학습 레이블 데이터 분포:\n', label_train.value_counts())
    print('검증 레이블 데이터 분포 :\n', label_test.value_counts())
    iris_df['label'].value_counts()

## 교차 검증 : 1
학습 레이블 데이터 분포:
 2    50
1    50
Name: label, dtype: int64
검증 레이블 데이터 분포 :
 0    50
Name: label, dtype: int64
## 교차 검증 : 2
학습 레이블 데이터 분포:
 2    50
0    50
Name: label, dtype: int64
검증 레이블 데이터 분포 :
 1    50
Name: label, dtype: int64
## 교차 검증 : 3
학습 레이블 데이터 분포:
 1    50
0    50
Name: label, dtype: int64
검증 레이블 데이터 분포 :
 2    50
Name: label, dtype: int64


- 아래에서 kfold 로 분할된 레이블 데이터 세트가 전체 레이블 값의 분포도를 반영하지 못하는 문제를 해결하기 위해 Stratifiedkfold 를 사용 할 것이다.
- 차이는 stratifiedKFold 는 레이블 데이터 분포도에 따라 학습/검증 데이터를 나누기 때문에 split() 매서드에 인자로 피처 데이터 세트 뿐만 아니라 레이블 데이터 세트도 반드시 필요하다는 사실임.

In [16]:
#StratifiedKFold 로 데이터 분할을 수행하고 학습/검증 레이블 데이터의 분포도를 확인.
from sklearn.model_selection import StratifiedKFold

skf = StratifiedKFold(n_splits=3)
n_iter=0

for train_index, test_index in skf.split(iris_df, iris_df['label']):
    n_iter+=1
    label_train=iris_df['label'].iloc[train_index]
    label_test=iris_df['label'].iloc[test_index]
    print('## 교차 검증: {0}'.format(n_iter))
    print('학습 레이블 데이터 분포:\n', label_train.value_counts())
    print('검증 레이블 데이터 분포:\n', label_test.value_counts())

## 교차 검증: 1
학습 레이블 데이터 분포:
 2    34
1    33
0    33
Name: label, dtype: int64
검증 레이블 데이터 분포:
 1    17
0    17
2    16
Name: label, dtype: int64
## 교차 검증: 2
학습 레이블 데이터 분포:
 1    34
2    33
0    33
Name: label, dtype: int64
검증 레이블 데이터 분포:
 2    17
0    17
1    16
Name: label, dtype: int64
## 교차 검증: 3
학습 레이블 데이터 분포:
 0    34
2    33
1    33
Name: label, dtype: int64
검증 레이블 데이터 분포:
 2    17
1    17
0    16
Name: label, dtype: int64


- 출력 결과를 보면 학습 렐이블과 검증 레이블 데이터 값의 분포도가 동일하게 할당되었음

In [17]:
# stratifiedKFold 를 이용해 데이터 분리 : 피처 데이터와 레이블 데이터는 앞의 붓꽃 stratifiedkfold 예제 추출 데이터 이용
dt_clf = DecisionTreeClassifier(random_state=156)

skfold = StratifiedKFold(n_splits=3)
n_iter=0
cv_accuracy=[]

#StratifiedKFold 의 split() 호출시 반드시 레이블 데이터 세트도 추가 입력 필요
for train_index, test_index in skfold.split(features, label):
    # split() 으로 반환된 인덱스를 이용해 학습용, 검증용 테스트 데이터 추출
    X_train, X_test = features[train_index], features[test_index]
    y_train, y_test = label[train_index], label[test_index]
    # 학습 및 예측
    dt_clf.fit(X_train, y_train)
    pred= dt_clf.predict(X_test)

    #반복 시마다 정확도 측정
    n_iter +=1
    accuracy = np.round(accuracy_score(y_test, pred),4)
    train_size = X_train.shape[0]
    test_size = X_test.shape[0]
    print('\n#{0} 교차 검증 정확도 :{1}, 학습 데이터 크기: {2}, 검증 데이터 크기: {3}'.format(n_iter, accuracy, train_size, test_size))
    print('#{0} 검증 세트 인덱스 :{1}'.format(n_iter, test_index))
    cv_accuracy.append(accuracy)

# 교차 검증별 정확도 및 평균 정확도 계산
print('\n## 교차 검증별 정확도:', np.round(cv_accuracy,4))
print('## 평균 검증 정확도 :', np.mean(cv_accuracy))


#1 교차 검증 정확도 :0.98, 학습 데이터 크기: 100, 검증 데이터 크기: 50
#1 검증 세트 인덱스 :[  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  50
  51  52  53  54  55  56  57  58  59  60  61  62  63  64  65  66 100 101
 102 103 104 105 106 107 108 109 110 111 112 113 114 115]

#2 교차 검증 정확도 :0.94, 학습 데이터 크기: 100, 검증 데이터 크기: 50
#2 검증 세트 인덱스 :[ 17  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  67
  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82 116 117 118
 119 120 121 122 123 124 125 126 127 128 129 130 131 132]

#3 교차 검증 정확도 :0.98, 학습 데이터 크기: 100, 검증 데이터 크기: 50
#3 검증 세트 인덱스 :[ 34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  83  84
  85  86  87  88  89  90  91  92  93  94  95  96  97  98  99 133 134 135
 136 137 138 139 140 141 142 143 144 145 146 147 148 149]

## 교차 검증별 정확도: [0.98 0.94 0.98]
## 평균 검증 정확도 : 0.9666666666666667


- 왜곡된 레이블 데이터 세트에서는 반드시 Stratified K 폴드를 이용해 교차 검증해야 함.
- 일반적으로 분류 에서의 교차 검증은 k 폴드가 아니라 stratified k 폴드로 분할돼야 함.
- 회귀 에서는 stratified k 폴드가 지원되지 않는다.

<교차 검증을 보다 간편하게 - cross_val_score()>
- 사이킷런은 교차 검증을 좀 더 편리하게 수행할 수 있게 해주는 API 를 제공한다.
- 그전에 KFold 로 데이터를 학습하고 예측하는 코드 복습
1. 폴드 세트를 선택하고
2. for 루프에서 반복으로 학습 및 테스트 데이터의 인덱스를 추출한 뒤
3. 반복적으로 학습과 예측을 하고 예측 성능을 반환
- cross_val_score() : 위의 일련의 과정을 한꺼번에 수행하는 API
- cross_vla_score(estimator, X, y=None, scoring = None, cv=None......)
- 반환값은 scoring 파라미터로 지정된 성능 지표 측정값을 배열 형태로 반환한다.
- cross_val_score() 은 classifier 가 입력되면 Stratified K 폴드 방식으로 레이블 값의 분포에 따라 학습/ 테스트 세트를 분할한다.

In [18]:
# corss_val_score() 사용법 알아보기. 교차 검증 폴드 수는 3, 성능평가 지표는 정확도인 accuracy로
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score, cross_validate
from sklearn.datasets import load_iris

iris_data= load_iris()
dt_clf=DecisionTreeClassifier(random_state=156)

data=iris_data.data
label=iris_data.target

# 성능 지표는 정확도(accuracy), 교차 검증 세트는 3개
scores = cross_val_score(dt_clf, data,label, scoring='accuracy', cv=3)
print('교차 검증별 정확도:', np.round(scores, 4))
print('평균 검증 정확도:', np.round(np.mean(scores),4))

교차 검증별 정확도: [0.98 0.94 0.98]
평균 검증 정확도: 0.9667


- cross_val_score() 은 cv로 지정된 횟수만큼 scoring 파라미터로 지정된 평가 지표로 평가 결괏값을 배열로 반환
- cross_val_score() 가 내부적으로 stratifiedkfold 를 이용하기 때문.

<GridSearchCV - 교차 검증과 최적 하이퍼 파라미터 튜닝을 한 번에>
- 하이퍼 파라미터는 머신러닝 알고리즘을 구성하는 주요 구성 요소이며, 이 값을 조정해 알고리즘의 예측 성능을 개선할 수 있다.
- 결정 트리 알고리즘의 여러 하이퍼 파라미터를 순차적으로 변경하면서 최고 성능을 가지는 파라미터 조합을 가지고자 한다면, 다음과 같이 파라미터의 집합을 만들고 이를 순차적으로 적용하면서 최적화를 수행할 수 있다.

In [19]:
grid_parameters = {'max_depth':[1,2,3],
                   'min_samples_split': [2,3]}

- GridSearchCV 는 사용자가 튜닝하고자 하는 여러 종류의 하이퍼 파라미터를 다양하게 테스트하면서 최적의 파라미터를 편리하게 찾게 해주지만, 동시에 순차적으로 파라미터를 테스트하므로 수행시간이 상대적으로 오래 걸린다.
- GridSearchCV 클래스의 생성자로 들어가는 주요 파라미터
1) estimator : classifier, regressor, pipeline 이 사용될 수 있다.
2) param_grid : key + 리스트 값을 가지는 딕셔너리가 주어진다.
3) scoring : 예측 성능을 측정할 평가 방법을 지정한다.
4) cv : 교차 검증을 위해 분할되는 학습/ 테스트 세트의 개수 지정
5) refit : 디폴트가 True 이며, 가장 최적의 하이퍼 파라미터를 찾은 뒤 입력된 estimator 객체를 해당 하이퍼 파라미터로 재학습시킨다.

In [20]:
# train_test_split() 을 이용해 학습 데이터와 테스트 데이터를 먼저 분리하고 학습 데이터에서 GridSearchCV를 이용해 최적 하이퍼 파라미터를 추출
# 테스트할 하이퍼 파라미터 세트는 딕셔너리 형태로 하이퍼 파라미터의 명칭은 문자열 Key 값으로, 하이퍼 파라미터의 값은 리스트 형으로 설정한다.
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

# 데이터를 로딩하고 학습 데이터와 텍스트 데이터 분리
iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris_data.data, iris_data.target, test_size=0.2, random_state=121)
dtree= DecisionTreeClassifier()

#파라미터를 딕셔너리 형태로 설정
parameters = {'max_depth':[1,2,3],
                   'min_samples_split': [2,3]}

- 학습된 데이터 세트를 ~~~~~~

In [21]:
import pandas as pd

# param_grid 의 하이퍼 파라미터를 3개의 train, test set fold로 나누어 테스트 수행 설정.
### refit=True 가 default 임. True이면 가장 좋은 파라미터 설정으로 재학습시킴.
grid_dtree=GridSearchCV(dtree, param_grid=parameters, cv=3, refit=True)

# 붓꽃 학습 데이터로 param_grid의 하이퍼 파라미터를 순차적으로 학습/평가
grid_dtree.fit(X_train, y_train)

# GridSearchCV 결과를 추출해 DataFrame 으로 변환
scores_df = pd.DataFrame(grid_dtree.cv_results_)
scores_df[['params', 'mean_test_score', 'rank_test_score', 'split0_test_score', 'split1_test_score', 'split2_test_score']]

params  mean_test_score  rank_test_score  \
0  {'max_depth': 1, 'min_samples_split': 2}         0.700000                5   
1  {'max_depth': 1, 'min_samples_split': 3}         0.700000                5   
2  {'max_depth': 2, 'min_samples_split': 2}         0.958333                3   
3  {'max_depth': 2, 'min_samples_split': 3}         0.958333                3   
4  {'max_depth': 3, 'min_samples_split': 2}         0.975000                1   
5  {'max_depth': 3, 'min_samples_split': 3}         0.975000                1   

   split0_test_score  split1_test_score  split2_test_score  
0              0.700                0.7               0.70  
1              0.700                0.7               0.70  
2              0.925                1.0               0.95  
3              0.925                1.0               0.95  
4              0.975                1.0               0.95  
5              0.975                1.0               0.95

- max_depth 와 min_samples_split 을 순차적으로 총 6번 변경하면서 학습 및 평가를 수행했음을 나타낸다.
- rank_test_score 을 보면 예측 성능이 1위 라는 의미
- 주요 칼럼별 의미
1) params : 칼럼에는 수행할 때 마다 적용된 개별 하이퍼 파라미터 값을 나타냄
2) rank_test_score : 하이퍼 파라미터별로 성능이 좋은 score 순위를 나타냄. 1이 가장 뛰어난 순위이며 이때의 파라미터가 최적의 하이퍼 파라미터이다.
3) mean_test_score : 개별 하이퍼 파라미터별로 CV의 폴딩 테스트 세트에 대해 총 수행한 평가 평균값

In [ ]:
print('GridSearchCV 최적 파라미터:', grid_dtree.best_params_)
print('GridSearchCV 최고 정확도:', format(grid_dtree.best_score_))

In [22]:
#GridSearchCV 의 refit 으로 이미 학습된 estimator 변환
estimator = grid_dtree.best_estimator_

#GridSearchCV의 best_estimator_ 는 이미 최적 학습이 됐으므로 별도 학습이 필요 없음
pred = estimator.predict(X_test)
print('테스트 데이터 세트 정확도: {0:4f}'.format(accuracy_score(y_test, pred)))

테스트 데이터 세트 정확도: 0.966667


- 일반적으로 학습 데이터를 GridSearchCV를 이용해 최적 하이퍼 파라미터 튜닝을 수행한 뒤에 별도의 테스트 세트에서 이를 평가하는 것이 일반적인 머신러닝 모델 적용 방법

5. 데이터 전처리
- 데이터 전처리는 ML 알고지름만큼 중요함.
- 데이터에 대해서 NULL 값등 미리 처리해야 하는 기본 사항

<데이터 인코딩>
- 레이블 인코딩과 원핫 인코딩이 있음.
- 레이블 인코팅 : 레이블 인코딩은 카테고리 피처를 코드형 숫자 값으로 변환하는 것

In [24]:
from sklearn.preprocessing import LabelEncoder

items = ['TV', '냉장고', '전자레인지', '컴퓨터', '선풍기', '선풍기', '믹서', '믹서']

# LabelEncoder 를 객체로 생성한 후, fit()과 transform() 으로 레이블 인코딩 수행.
encoder = LabelEncoder()
encoder.fit(items)
labels=encoder.transform(items)
print('인코딩 변환값:', labels)

인코딩 변환값: [0 1 4 5 3 3 2 2]


In [25]:
print('인코딩 클래스:', encoder.classes_)

인코딩 클래스: ['TV' '냉장고' '믹서' '선풍기' '전자레인지' '컴퓨터']
